### ISSUE TITLE
multi-stage feedback

### ISSUE THREAD

**Issue by mantoni** [npm/npm#8515]
First of all, thank you @iarna for all the hard work. It solves our main issue, which is install / dedupe times of ~15 minutes. The multi-stage feature brings this down to ~90 seconds. __This is fantastic!__

There are a few findings that I'd like to share. I don't know how much of this you aleady know, so feel free to close right away.

* I checked out the `multi-stage` branch and installed with `node cli.js install -g -f`. This causes some issues with our company proxy (which also makes `npm i npm/npm#multi-stage` impossible). It would be fantastic if we could have a tagged npm version in the registry. This would allow my team to become beta-testers for multi-stage.
* `npm install` does not install devDependencies anymore. Is this intentional? How do I get it to install them? This seems to be the only blocker for me at the moment.
* As I understand, `npm ls` is now showing the logical dependencies instead of the "physical". If I have two top level modules that both require the same version of another module, it get's installed only once correctly, but `npm ls` only shows it as a dependency of the first module. I'd expect it to be either top level (as it's in the file system) or listed twice (maybe with a label stating that this is not the physical location of the dependency?)
* `npm dedupe` currently produces this error (I can only reproduce on my project):

```
npm ERR! Linux 2.6.18-308.el5
npm ERR! argv "node" "/usr/bin/npm" "dedupe"
npm ERR! node v0.10.38
npm ERR! npm  v3.0.0-pre.11
npm ERR! path /home/.../projects/.../node_modules/express/node_modules/send/node_modules/ms
npm ERR! code EACCES

npm ERR! Error: EACCES, access '/home/.../projects/.../node_modules/express/node_modules/send/node_modules/ms'
npm ERR!     at accessError (/home/.../npm/lib/install/check-permissions.js:88:18)
npm ERR!     at /home/.../npm/lib/install/check-permissions.js:79:29
npm ERR!     at OpenReq.Req.done (/home/.../npm/node_modules/graceful-fs/graceful-fs.js:141:5)
npm ERR!     at OpenReq.done (/home/.../npm/node_modules/graceful-fs/graceful-fs.js:61:22)
npm ERR!     at Object.oncomplete (evalmachine.<anonymous>:108:15)
npm ERR!  { [Error: EACCES, access '/home/.../projects/.../node_modules/express/node_modules/send/node_modules/ms']
npm ERR!   code: 'EACCES',
npm ERR!   path: '/home/.../projects/.../node_modules/express/node_modules/send/node_modules/ms' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.
```

Obviously, dedupe is not so much of an issue anymore because the tree is already dedupe on install :+1:

Other than the obove, it works fine for me. If there was a tagged preview release in the registry, I'd start using it regularly and report bugs. Let me know if it's too early for this and I'll hold my horses.

/cc @othiym23 @zkat

**Comment by mantoni** [comment]
Update: devDependencies occur in the progress bar, but don't show up in the tree after installing. I fixed it locally by removing this block: https://github.com/npm/npm/blob/multi-stage/lib/install/prune-tree.js#L27-L34

**Comment by bengl** [comment]
> `npm install` does not install devDependencies anymore.

and

> Update: devDependencies occur in the progress bar, but don't show up in the tree after installing.

also #8525

**Comment by iarna** [comment]
@mantoni

devDependencies were getting resolved and then removed by the pruner. The pruner was... a touch overzealous. 😉

> I checked out the multi-stage branch and installed with node cli.js install -g -f. This causes some issues with our company proxy (which also makes npm i npm/npm#multi-stage impossible). It would be fantastic if we could have a tagged npm version in the registry. This would allow my team to become beta-testers for multi-stage.

We're not going to do this until we've finished until we've eliminated the known issues with it. There definitely will be a period of time with tagged 3.0 releases that aren't `next` or `latest` though, and we would invite your team to test then! 😄

> npm install does not install devDependencies anymore. Is this intentional? How do I get it to install them? This seems to be the only blocker for me at the moment.

That was a bug in the pruner. They were being added to the virtual tree and then removed as unneeded. 😒It's been resolved now if you want to try out this week's version.

> As I understand, npm ls is now showing the logical dependencies instead of the "physical". If I have two top level modules that both require the same version of another module, it get's installed only once correctly, but npm ls only shows it as a dependency of the first module. I'd expect it to be either top level (as it's in the file system) or listed twice (maybe with a label stating that this is not the physical location of the dependency?)

It should be listed twice, but I believe there's some legacy code kicking around in `npm ls` that tries to prune the output. Before all is said and done, we'll want to revisit that. Personally, I'd also like some way to see the physical layout of the tree.

**Comment by iarna** [comment]
> npm dedupe currently produces this error (I can only reproduce on my project)

Hey neat, you ran afoul of the new permissions checker! So we DEFINITELY need some better messaging around that, right now you're just getting the raw error object dumped out.

What it's trying to tell you is that it needs to write to `/home/.../projects/.../node_modules/express/node_modules/send/node_modules/ms` and it can't.

**Comment by mantoni** [comment]
Thanks for all the explanations. Your error checkers seems to not like what npm just created then ;-) The path it's complaining about was created during a fresh install.

**Comment by iarna** [comment]
mantoni: I can't reproduce that. It would be super helpful if you could check the permissions on those paths?

**Comment by iarna** [comment]
(Don't get me wrong: I'm not suggesting that you broke something in your install, but diagnostically, if `npm install` is mucking up permissions somehow it would be _really_ helpful to see how.)

**Comment by mantoni** [comment]
Sure. No worries. Will do tomorrow. It's a little late over here right now ;-)

**Comment by iarna** [comment]
Thanks! 😄

**Comment by mantoni** [comment]
Got it. There is nothing wrong with the file permissions. It's trying to access a file that isn't there. Let me explain why:

- We use an `npm-shrinkwrap.json` file
- I did an `npm install --no-shrinkwrap` which went through fine
- Next I ran `npm dedupe` which yielded the error because npm seems to take the path where it expects the file from the  `npm-shrinkwrap.json`.
- After finding this, I tried `npm dedupe --no-shrinkwrap` and that passed without error (and without doing any modifications to the already flat tree)

For us this will be a academical problem because we will stop using dedupe once we switched to npm 3. Our build process always creates fresh installs by deleting node_modules. However, one could argue that `npm dedupe` should never make assumptions about the existing tree structure from looking into the shrinkwrap. Your call.

**Comment by mantoni** [comment]
I did some further testing on the current `multi-stage` and found that all the issues listed above have been addressed (apart from the dedupe thingy, which might not even be a bug).  :tada:
 
Thanks @iarna, leaving it to you to close the issue.

**Comment by iarna** [comment]
Ah-hah! Yeah, dedupe is now built on the same code as the installer, so it's loading the shrinkwrap when it's doing it initially figures out what the ideal tree. But that's definitely not right! We want to dedupe the tree you have, not the tree you want to have. 😉

**Comment by iarna** [comment]
And this is fixed with dcfaf6c
